File: trabalho1/src/interpreter_graph/entity/matrix_adjacent.py
from __future__ import annotations

import logging
from typing import List

log = logging.getLogger(__name__)


class MatrixAdjacent:
    _total_vertexes: int
    _matrix_adjacent: List[List[int]]

    def __new__(cls, *args, **kwargs):
        return super().__new__(cls)

    def __init__(self, total_vertexes: int):
        if total_vertexes < 1:
            raise RuntimeError("Vertices should be more than 1")

        self._total_vertexes = total_vertexes
        self._matrix_adjacent = [[0] * total_vertexes] * total_vertexes

    def _validate_bounds_vertex(self, vertex: int) -> bool:
        return 0 <= vertex < self._total_vertexes

    def _validate_vertexes_to_edge(self, vertex1: int, vertex2: int) -> None:
        log.debug(f"validating vertexes, entry ({vertex1}, {vertex2})")
        is_valid_vertex1 = self._validate_bounds_vertex(vertex1)
        is_valid_vertex2 = self._validate_bounds_vertex(vertex2)
        if (not is_valid_vertex1) or (not is_valid_vertex2):
            message_error = f"Bounds of vertex shound be major than 0 and minour than {self._total_vertexes}, the entry value was ({vertex1},{vertex2})"
            raise RuntimeError(message_error)

    def add_edge(self, vertex1: int, vertex2: int) -> MatrixAdjacent:
        self._validate_vertexes_to_edge(vertex1, vertex2)

        self._matrix_adjacent[vertex1][vertex2] = 1
        self._matrix_adjacent[vertex2][vertex1] = 1
        return self

    def print_matrix(self) -> None:
        for lines in self._matrix_adjacent:
            print(lines)

File: trabalho1-kotlin/src/main/kotlin/aragao/ellian/github/graph/entity/MatrixAdjacent.kt
package aragao.ellian.github.graph.entity

class MatrixAdjacent(private val totalVertex: Int, private var preCompute: Boolean = false) {
    private val matrixAdjacent = Array(totalVertex) { IntArray(totalVertex) }
    private var vertexDegree = Array(totalVertex) { 0 }
    private var edgeLenght = 0L

    private fun validateBoundsVertex(vertex: Int): Boolean = 0 <= vertex && vertex < this.totalVertex

    private fun validateVertexesToEdge(vertex1: Int, vertex2: Int) {
        val isValidVertex1 = validateBoundsVertex(vertex1)
        val isValidVertex2 = validateBoundsVertex(vertex2)
        if (!isValidVertex1 or !isValidVertex2) {
            val message =
                "Bounds of vertex shound be major than 0 and minour than ${this.totalVertex}, the entry value was (${vertex1}, ${vertex2})"
            throw RuntimeException(message)
        }
    }

    fun addEdge(vertex1: Int, vertex2: Int): MatrixAdjacent {
        validateVertexesToEdge(vertex1, vertex2)
        matrixAdjacent[vertex1][vertex2]++
        matrixAdjacent[vertex2][vertex1]++
        edgeLenght++
        if (preCompute)
            processBiVertexDegree(vertex1, vertex2)
        return this
    }

    private fun processBiVertexDegree(vertex1: Int, vertex2: Int) {
        vertexDegree[vertex1] = processVertexDegree(matrixAdjacent[vertex1])
        vertexDegree[vertex2] = processVertexDegree(matrixAdjacent[vertex2])
    }

    private fun processVertexDegree(vertexLine: IntArray): Int = vertexLine.reduce(Int::plus)

    fun processAllVertexesDegrees(): MatrixAdjacent {
        vertexDegree = matrixAdjacent
            .map(this::processVertexDegree)
            .toTypedArray()
        return this
    }

    override fun toString(): String {
        return "MatrixAdjacent(${System.lineSeparator()}" +
                " totalVertex = $totalVertex, ${System.lineSeparator()}" +
                " preCompute = $preCompute, ${System.lineSeparator()}" +
                " edgeLenght = $edgeLenght,${System.lineSeparator()}" +
                " vertexDegree: ${System.lineSeparator()}  ${vertexDegree.contentToString()}${System.lineSeparator()}" +
                " matrixAdjacent: ${System.lineSeparator()}  " +
                matrixAdjacent.joinToString("  ") { "[${it.joinToString()}]${System.lineSeparator()}" } +
                ")"
    }

    fun getVertexDegress() = vertexDegree
    fun getEdgeLenght() = edgeLenght
}
